File: src/templates/claude/hooks/session-start.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Session Start Hook - Inject structured context."""

from __future__ import annotations

import warnings

warnings.filterwarnings("ignore")

import json
import os
import subprocess
import sys
from io import StringIO
from pathlib import Path

if sys.platform == "win32":
    import io as _io

    if hasattr(sys.stdout, "reconfigure"):
        sys.stdout.reconfigure(encoding="utf-8", errors="replace")  # type: ignore[union-attr]
    elif hasattr(sys.stdout, "detach"):
        sys.stdout = _io.TextIOWrapper(sys.stdout.detach(), encoding="utf-8", errors="replace")  # type: ignore[union-attr]


def should_skip_injection() -> bool:
    return (
        os.environ.get("CLAUDE_NON_INTERACTIVE") == "1"
        or os.environ.get("OPENCODE_NON_INTERACTIVE") == "1"
    )


def read_file(path: Path, fallback: str = "") -> str:
    try:
        return path.read_text(encoding="utf-8")
    except (FileNotFoundError, PermissionError):
        return fallback


def run_script(script_path: Path, *args: str) -> str:
    if not script_path.is_file():
        return ""

    try:
        env = os.environ.copy()
        env["PYTHONIOENCODING"] = "utf-8"
        cmd = [sys.executable, "-W", "ignore", str(script_path), *args]
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=5,
            cwd=script_path.parent.parent.parent,
            env=env,
        )
        return result.stdout if result.returncode == 0 else ""
    except (subprocess.TimeoutExpired, FileNotFoundError, PermissionError):
        return ""


def _normalize_task_ref(task_ref: str) -> str:
    normalized = task_ref.strip()
    if not normalized:
        return ""

    path_obj = Path(normalized)
    if path_obj.is_absolute():
        return str(path_obj)

    normalized = normalized.replace("\\", "/")
    while normalized.startswith("./"):
        normalized = normalized[2:]

    if normalized.startswith("tasks/"):
        return f".trellis/{normalized}"

    return normalized


def _resolve_task_dir(trellis_dir: Path, task_ref: str) -> Path:
    normalized = _normalize_task_ref(task_ref)
    path_obj = Path(normalized)
    if path_obj.is_absolute():
        return path_obj
    if normalized.startswith(".trellis/"):
        return trellis_dir.parent / path_obj
    return trellis_dir / "tasks" / path_obj


def _get_task_status(trellis_dir: Path) -> str:
    current_task_file = trellis_dir / ".current-task"
    if not current_task_file.is_file():
        return "Status: NO ACTIVE TASK\nNext: Describe what you want to work on"

    task_ref = _normalize_task_ref(current_task_file.read_text(encoding="utf-8").strip())
    if not task_ref:
        return "Status: NO ACTIVE TASK\nNext: Describe what you want to work on"

    task_dir = _resolve_task_dir(trellis_dir, task_ref)
    if not task_dir.is_dir():
        return f"Status: STALE POINTER\nTask: {task_ref}\nNext: Task directory not found. Run: python3 ./.trellis/scripts/task.py finish"

    task_json_path = task_dir / "task.json"
    task_data = {}
    if task_json_path.is_file():
        try:
            task_data = json.loads(task_json_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, PermissionError):
            pass

    task_title = task_data.get("title", task_ref)
    task_status = task_data.get("status", "unknown")

    if task_status == "completed":
        return f"Status: COMPLETED\nTask: {task_title}\nNext: Archive with `python3 ./.trellis/scripts/task.py archive {task_dir.name}` or start a new task"

    has_context = False
    for jsonl_name in ("implement.jsonl", "check.jsonl", "spec.jsonl"):
        jsonl_path = task_dir / jsonl_name
        if jsonl_path.is_file() and jsonl_path.stat().st_size > 0:
            has_context = True
            break

    has_prd = (task_dir / "prd.md").is_file()

    if not has_prd:
        return f"Status: NOT READY\nTask: {task_title}\nMissing: prd.md not created\nNext: Write PRD, then research → init-context → start"

    if not has_context:
        return f"Status: NOT READY\nTask: {task_title}\nMissing: Context not configured (no jsonl files)\nNext: Complete Phase 2 (research → init-context → start) before implementing"

    return f"Status: READY\nTask: {task_title}\nNext: Continue with implement or check"


def _load_trellis_config(trellis_dir: Path) -> tuple[bool, dict, object, str | None, str | None]:
    scripts_dir = trellis_dir / "scripts"
    if str(scripts_dir) not in sys.path:
        sys.path.insert(0, str(scripts_dir))

    try:
        from common.config import get_default_package, get_packages, get_spec_scope, is_monorepo  # type: ignore[import-not-found]
        from common.paths import get_current_task  # type: ignore[import-not-found]

        repo_root = trellis_dir.parent
        is_mono = is_monorepo(repo_root)
        packages = get_packages(repo_root) or {}
        scope = get_spec_scope(repo_root)

        task_pkg = None
        current = get_current_task(repo_root)
        if current:
            task_json = repo_root / current / "task.json"
            if task_json.is_file():
                try:
                    data = json.loads(task_json.read_text(encoding="utf-8"))
                    if isinstance(data, dict):
                        candidate = data.get("package")
                        if isinstance(candidate, str) and candidate:
                            task_pkg = candidate
                except (json.JSONDecodeError, OSError):
                    pass

        default_pkg = get_default_package(repo_root)
        return is_mono, packages, scope, task_pkg, default_pkg
    except Exception:
        return False, {}, None, None, None


def _check_legacy_spec(trellis_dir: Path, is_mono: bool, packages: dict) -> str | None:
    if not is_mono or not packages:
        return None

    spec_dir = trellis_dir / "spec"
    if not spec_dir.is_dir():
        return None

    has_legacy = False
    for legacy_name in ("backend", "frontend"):
        legacy_dir = spec_dir / legacy_name
        if legacy_dir.is_dir() and (legacy_dir / "index.md").is_file():
            has_legacy = True
            break

    if not has_legacy:
        return None

    missing = [name for name in sorted(packages.keys()) if not (spec_dir / name).is_dir()]
    if not missing:
        return None

    if len(missing) == len(packages):
        return (
            "[!] Legacy spec structure detected: found `spec/backend/` or `spec/frontend/` "
            "but no package-scoped `spec/<package>/` directories.\n"
            f"Monorepo packages: {', '.join(sorted(packages.keys()))}\n"
            "Please reorganize: `spec/backend/` -> `spec/<package>/backend/`"
        )

    return (
        f"[!] Partial spec migration detected: packages {', '.join(missing)} "
        "still missing `spec/<pkg>/` directory.\n"
        "Please complete migration for all packages."
    )


def _resolve_spec_scope(
    is_mono: bool,
    packages: dict,
    scope: object,
    task_pkg: str | None,
    default_pkg: str | None,
) -> set[str] | None:
    if not is_mono or not packages:
        return None

    if scope is None:
        return None

    if isinstance(scope, str) and scope == "active_task":
        if task_pkg and task_pkg in packages:
            return {task_pkg}
        if default_pkg and default_pkg in packages:
            return {default_pkg}
        return None

    if isinstance(scope, list):
        valid = {entry for entry in scope if entry in packages}
        if valid:
            return valid
        if task_pkg and task_pkg in packages:
            return {task_pkg}
        if default_pkg and default_pkg in packages:
            return {default_pkg}
        return None

    return None


def _build_workflow_toc(workflow_path: Path) -> str:
    content = read_file(workflow_path)
    if not content:
        return "No workflow.md found"

    toc_lines = [
        "# Development Workflow — Section Index",
        "Full guide: .trellis/workflow.md (read on demand)",
        "",
    ]
    for line in content.splitlines():
        if line.startswith("## "):
            toc_lines.append(line)

    toc_lines += [
        "",
        "To read a section: use the Read tool on .trellis/workflow.md",
    ]
    return "\n".join(toc_lines)


def _load_session_gate_summary(trellis_dir: Path) -> str | None:
    gate_script = trellis_dir / "scripts" / "session_gate.py"
    raw = run_script(gate_script, "show", "--json")
    if not raw:
        return None

    try:
        state = json.loads(raw)
    except json.JSONDecodeError:
        return None

    status = str(state.get("status", "not_evaluated"))
    if status == "not_evaluated":
        return None

    lines = [f"Status: {status.upper()}"]

    task_type = state.get("taskType")
    if task_type:
        lines.append(f"Task type: {task_type}")

    request = state.get("request")
    if request:
        lines.append(f"Request: {request}")

    summary = state.get("summary")
    if summary:
        lines.append(f"Summary: {summary}")

    docs = state.get("applicableDocs") or []
    if docs:
        lines.append("Docs:")
        for item in docs:
            lines.append(f"- {item}")

    blockers = state.get("blockers") or []
    if blockers:
        lines.append("Blockers:")
        for blocker in blockers:
            lines.append(f"- {blocker}")

    next_step = state.get("nextStep")
    if next_step:
        lines.append(f"Next: {next_step}")

    updated_at = state.get("updatedAt")
    if updated_at:
        lines.append(f"Updated: {updated_at}")

    return "\n".join(lines)


def _append_index(output: StringIO, title: str, index_file: Path) -> bool:
    content = read_file(index_file)
    if not content:
        return False
    output.write(f"## {title}\n")
    output.write(content)
    output.write("\n\n")
    return True


def _write_guideline_indexes(output: StringIO, spec_dir: Path, allowed_pkgs: set[str] | None) -> None:
    has_content = False
    if not spec_dir.is_dir():
        output.write("Not configured")
        return

    root_index = spec_dir / "index.md"
    if root_index.is_file():
        has_content = _append_index(output, "spec", root_index) or has_content

    guides_index = spec_dir / "guides" / "index.md"
    if guides_index.is_file():
        has_content = _append_index(output, "guides", guides_index) or has_content

    for top_level in sorted(spec_dir.iterdir(), key=lambda path: path.name):
        if not top_level.is_dir() or top_level.name.startswith(".") or top_level.name == "guides":
            continue

        top_index = top_level / "index.md"
        if top_index.is_file():
            has_content = _append_index(output, top_level.name, top_index) or has_content
            for child in sorted(top_level.iterdir(), key=lambda path: path.name):
                if not child.is_dir() or child.name.startswith("."):
                    continue
                child_index = child / "index.md"
                if child_index.is_file():
                    has_content = _append_index(output, f"{top_level.name}/{child.name}", child_index) or has_content
            continue

        if allowed_pkgs is not None and top_level.name not in allowed_pkgs:
            continue

        for child in sorted(top_level.iterdir(), key=lambda path: path.name):
            if not child.is_dir() or child.name.startswith("."):
                continue
            child_index = child / "index.md"
            if child_index.is_file():
                has_content = _append_index(output, f"{top_level.name}/{child.name}", child_index) or has_content

    if not has_content:
        output.write("Not configured")


def main() -> None:
    if should_skip_injection():
        sys.exit(0)

    project_dir = Path(os.environ.get("CLAUDE_PROJECT_DIR", ".")).resolve()
    trellis_dir = project_dir / ".trellis"

    is_mono, packages, scope_config, task_pkg, default_pkg = _load_trellis_config(trellis_dir)
    allowed_pkgs = _resolve_spec_scope(is_mono, packages, scope_config, task_pkg, default_pkg)

    output = StringIO()
    output.write(
        """<session-context>
You are starting a new session in a Trellis-managed project.
Read and follow all instructions below carefully.
</session-context>

"""
    )

    legacy_warning = _check_legacy_spec(trellis_dir, is_mono, packages)
    if legacy_warning:
        output.write(f"<migration-warning>\n{legacy_warning}\n</migration-warning>\n\n")

    context_script = trellis_dir / "scripts" / "get_context.py"
    context_output = run_script(context_script)
    if context_output:
        output.write("<current-state>\n")
        output.write(context_output.rstrip())
        output.write("\n</current-state>\n\n")

    gate_summary = _load_session_gate_summary(trellis_dir)
    if gate_summary:
        output.write("<session-gate>\n")
        output.write(gate_summary)
        output.write("\n</session-gate>\n\n")

    output.write("<workflow>\n")
    output.write(_build_workflow_toc(trellis_dir / "workflow.md"))
    output.write("\n</workflow>\n\n")

    output.write("<guidelines>\n")
    output.write(
        "These are guideline indexes only. Read the specific files they reference before implementation.\n\n"
    )
    _write_guideline_indexes(output, trellis_dir / "spec", allowed_pkgs)
    output.write("\n</guidelines>\n\n")

    task_status = _get_task_status(trellis_dir)
    output.write(f"<task-status>\n{task_status}\n</task-status>\n\n")

    output.write(
        """<ready>
Context loaded. Workflow index, project state, guideline indexes, and any persisted session gate state are already injected above.
Wait for the user's first message, then follow the workflow.
If there is an active task, ask whether to continue it.
</ready>"""
    )

    result = {
        "hookSpecificOutput": {
            "hookEventName": "SessionStart",
            "additionalContext": output.getvalue(),
        }
    }
    print(json.dumps(result, ensure_ascii=False), flush=True)


if __name__ == "__main__":
    main()

File: src/templates/opencode/plugin/session-start.js
/* global process */
/**
 * Trellis Session Start Plugin
 *
 * Injects context when user sends the first message in a session.
 * Uses OpenCode's chat.message + experimental.chat.messages.transform hooks.
 *
 * Compatibility:
 * - If oh-my-opencode handles via .claude/hooks/, this plugin skips
 * - Otherwise, this plugin handles injection
 */

import { execFileSync } from "child_process"
import { existsSync, readdirSync } from "fs"
import { platform } from "os"
import { join } from "path"
import { TrellisContext, contextCollector, debugLog } from "../lib/trellis-context.js"

const PYTHON_CMD = platform() === "win32" ? "python" : "python3"

function normalizeTaskRef(taskRef) {
  const trimmed = String(taskRef || "").trim()
  if (!trimmed) return ""

  if (trimmed.startsWith("/") || /^[A-Za-z]:[\\/]/.test(trimmed)) {
    return trimmed
  }

  let normalized = trimmed.replace(/\\/g, "/")
  while (normalized.startsWith("./")) {
    normalized = normalized.slice(2)
  }
  if (normalized.startsWith("tasks/")) {
    return `.trellis/${normalized}`
  }
  return normalized
}

function resolveTaskDir(ctx, taskRef) {
  const normalized = normalizeTaskRef(taskRef)
  if (!normalized) return null
  if (normalized.startsWith("/") || /^[A-Za-z]:[\\/]/.test(normalized)) {
    return normalized
  }
  if (normalized.startsWith(".trellis/")) {
    return join(ctx.directory, normalized)
  }
  return join(ctx.directory, ".trellis", "tasks", normalized)
}

function getTaskStatus(ctx) {
  const taskRef = ctx.getCurrentTask()
  if (!taskRef) {
    return "Status: NO ACTIVE TASK\nNext: Describe what you want to work on"
  }

  const taskDir = resolveTaskDir(ctx, taskRef)
  if (!taskDir || !existsSync(taskDir)) {
    return `Status: STALE POINTER\nTask: ${taskRef}\nNext: Task directory not found. Run: python3 ./.trellis/scripts/task.py finish`
  }

  let taskData = {}
  const taskJsonPath = join(taskDir, "task.json")
  const taskJson = ctx.readFile(taskJsonPath)
  if (taskJson) {
    try {
      taskData = JSON.parse(taskJson)
    } catch {
      // ignore
    }
  }

  const taskTitle = taskData.title || normalizeTaskRef(taskRef)
  const taskState = taskData.status || "unknown"

  if (taskState === "completed") {
    const dirName = taskDir.split(/[\\/]/).pop()
    return `Status: COMPLETED\nTask: ${taskTitle}\nNext: Archive with \`python3 ./.trellis/scripts/task.py archive ${dirName}\` or start a new task`
  }

  let hasContext = false
  for (const jsonlName of ["implement.jsonl", "check.jsonl", "spec.jsonl"]) {
    const content = ctx.readFile(join(taskDir, jsonlName))
    if (content && content.trim()) {
      hasContext = true
      break
    }
  }

  const hasPrd = !!ctx.readFile(join(taskDir, "prd.md"))
  if (!hasPrd) {
    return `Status: NOT READY\nTask: ${taskTitle}\nMissing: prd.md not created\nNext: Write PRD, then research → init-context → start`
  }
  if (!hasContext) {
    return `Status: NOT READY\nTask: ${taskTitle}\nMissing: Context not configured (no jsonl files)\nNext: Complete Phase 2 (research → init-context → start) before implementing`
  }

  return `Status: READY\nTask: ${taskTitle}\nNext: Continue with implement or check`
}

function loadTrellisConfig(ctx) {
  const getContextScript = join(ctx.directory, ".trellis", "scripts", "get_context.py")
  if (!existsSync(getContextScript)) {
    return { isMonorepo: false, packages: {}, specScope: null, activeTaskPackage: null, defaultPackage: null }
  }

  try {
    const result = execFileSync(PYTHON_CMD, [getContextScript, "--mode", "packages", "--json"], {
      cwd: ctx.directory,
      timeout: 5000,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    })
    const data = JSON.parse(result)
    if (data.mode !== "monorepo") {
      return { isMonorepo: false, packages: {}, specScope: null, activeTaskPackage: null, defaultPackage: null }
    }
    const packages = {}
    for (const pkg of (data.packages || [])) {
      packages[pkg.name] = pkg
    }
    return {
      isMonorepo: true,
      packages,
      specScope: data.specScope || null,
      activeTaskPackage: data.activeTaskPackage || null,
      defaultPackage: data.defaultPackage || null,
    }
  } catch {
    return { isMonorepo: false, packages: {}, specScope: null, activeTaskPackage: null, defaultPackage: null }
  }
}

function checkLegacySpec(ctx, config) {
  if (!config.isMonorepo || Object.keys(config.packages).length === 0) {
    return null
  }

  const specDir = join(ctx.directory, ".trellis", "spec")
  if (!existsSync(specDir)) {
    return null
  }

  const hasLegacy = ["backend", "frontend"].some((name) => existsSync(join(specDir, name, "index.md")))
  if (!hasLegacy) {
    return null
  }

  const packageNames = Object.keys(config.packages).sort()
  const missing = packageNames.filter((name) => !existsSync(join(specDir, name)))
  if (missing.length === 0) {
    return null
  }

  if (missing.length === packageNames.length) {
    return [
      "[!] Legacy spec structure detected: found `spec/backend/` or `spec/frontend/` but no package-scoped `spec/<package>/` directories.",
      `Monorepo packages: ${packageNames.join(", ")}`,
      "Please reorganize: `spec/backend/` -> `spec/<package>/backend/`",
    ].join("\n")
  }

  return [
    `[!] Partial spec migration detected: packages ${missing.join(", ")} still missing \`spec/<pkg>/\` directory.`,
    "Please complete migration for all packages.",
  ].join("\n")
}

function resolveSpecScope(config) {
  if (!config.isMonorepo || Object.keys(config.packages).length === 0) {
    return null
  }

  const { specScope, activeTaskPackage, defaultPackage, packages } = config
  if (specScope == null) return null

  if (specScope === "active_task") {
    if (activeTaskPackage && activeTaskPackage in packages) return new Set([activeTaskPackage])
    if (defaultPackage && defaultPackage in packages) return new Set([defaultPackage])
    return null
  }

  if (Array.isArray(specScope)) {
    const valid = new Set(specScope.filter((entry) => entry in packages))
    if (valid.size > 0) return valid
    if (activeTaskPackage && activeTaskPackage in packages) return new Set([activeTaskPackage])
    if (defaultPackage && defaultPackage in packages) return new Set([defaultPackage])
  }

  return null
}

function buildWorkflowToc(workflow) {
  if (!workflow) {
    return "No workflow.md found"
  }

  const lines = [
    "# Development Workflow — Section Index",
    "Full guide: .trellis/workflow.md (read on demand)",
    "",
  ]

  for (const line of workflow.split(/\r?\n/)) {
    if (line.startsWith("## ")) {
      lines.push(line)
    }
  }

  lines.push("", "To read a section: use the Read tool on .trellis/workflow.md")
  return lines.join("\n")
}

function loadSessionGateSummary(ctx) {
  const gateScript = join(ctx.directory, ".trellis", "scripts", "session_gate.py")
  if (!existsSync(gateScript)) {
    return null
  }

  try {
    const raw = execFileSync(PYTHON_CMD, [gateScript, "show", "--json"], {
      cwd: ctx.directory,
      timeout: 5000,
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    })
    const state = JSON.parse(raw)
    const status = String(state.status || "not_evaluated")
    if (status === "not_evaluated") {
      return null
    }

    const lines = [`Status: ${status.toUpperCase()}`]
    if (state.taskType) lines.push(`Task type: ${state.taskType}`)
    if (state.request) lines.push(`Request: ${state.request}`)
    if (state.summary) lines.push(`Summary: ${state.summary}`)
    if (Array.isArray(state.applicableDocs) && state.applicableDocs.length > 0) {
      lines.push("Docs:")
      for (const item of state.applicableDocs) {
        lines.push(`- ${item}`)
      }
    }
    if (Array.isArray(state.blockers) && state.blockers.length > 0) {
      lines.push("Blockers:")
      for (const item of state.blockers) {
        lines.push(`- ${item}`)
      }
    }
    if (state.nextStep) lines.push(`Next: ${state.nextStep}`)
    if (state.updatedAt) lines.push(`Updated: ${state.updatedAt}`)
    return lines.join("\n")
  } catch {
    return null
  }
}

function appendIndex(parts, title, content) {
  if (!content) return false
  parts.push(`## ${title}\n${content}`)
  return true
}

function writeGuidelineIndexes(parts, ctx, allowedPkgs) {
  const specDir = join(ctx.directory, ".trellis", "spec")
  let hasContent = false

  if (!existsSync(specDir)) {
    parts.push("Not configured")
    return
  }

  const rootIndex = ctx.readFile(join(specDir, "index.md"))
  if (rootIndex) {
    hasContent = appendIndex(parts, "spec", rootIndex) || hasContent
  }

  const guidesIndex = ctx.readFile(join(specDir, "guides", "index.md"))
  if (guidesIndex) {
    hasContent = appendIndex(parts, "guides", guidesIndex) || hasContent
  }

  let entries = []
  try {
    entries = readdirSync(specDir, { withFileTypes: true })
  } catch {
    entries = []
  }

  for (const entry of entries.sort((a, b) => a.name.localeCompare(b.name))) {
    if (!entry.isDirectory() || entry.name.startsWith(".") || entry.name === "guides") {
      continue
    }

    const topName = entry.name
    const topIndex = ctx.readFile(join(specDir, topName, "index.md"))
    if (topIndex) {
      hasContent = appendIndex(parts, topName, topIndex) || hasContent

      let childEntries = []
      try {
        childEntries = readdirSync(join(specDir, topName), { withFileTypes: true })
      } catch {
        childEntries = []
      }

      for (const child of childEntries.sort((a, b) => a.name.localeCompare(b.name))) {
        if (!child.isDirectory() || child.name.startsWith(".")) {
          continue
        }
        const childIndex = ctx.readFile(join(specDir, topName, child.name, "index.md"))
        if (childIndex) {
          hasContent = appendIndex(parts, `${topName}/${child.name}`, childIndex) || hasContent
        }
      }
      continue
    }

    if (allowedPkgs !== null && !allowedPkgs.has(topName)) {
      continue
    }

    let childEntries = []
    try {
      childEntries = readdirSync(join(specDir, topName), { withFileTypes: true })
    } catch {
      childEntries = []
    }

    for (const child of childEntries.sort((a, b) => a.name.localeCompare(b.name))) {
      if (!child.isDirectory() || child.name.startsWith(".")) {
        continue
      }
      const childIndex = ctx.readFile(join(specDir, topName, child.name, "index.md"))
      if (childIndex) {
        hasContent = appendIndex(parts, `${topName}/${child.name}`, childIndex) || hasContent
      }
    }
  }

  if (!hasContent) {
    parts.push("Not configured")
  }
}

function buildSessionContext(ctx) {
  const config = loadTrellisConfig(ctx)
  const allowedPkgs = resolveSpecScope(config)
  const parts = []

  parts.push(`<session-context>\nYou are starting a new session in a Trellis-managed project.\nRead and follow all instructions below carefully.\n</session-context>`)

  const legacyWarning = checkLegacySpec(ctx, config)
  if (legacyWarning) {
    parts.push(`<migration-warning>\n${legacyWarning}\n</migration-warning>`)
  }

  const contextScript = join(ctx.directory, ".trellis", "scripts", "get_context.py")
  if (existsSync(contextScript)) {
    const output = ctx.runScript(contextScript)
    if (output) {
      parts.push(`<current-state>\n${output.trimEnd()}\n</current-state>`)
    }
  }

  const gateSummary = loadSessionGateSummary(ctx)
  if (gateSummary) {
    parts.push(`<session-gate>\n${gateSummary}\n</session-gate>`)
  }

  parts.push(`<workflow>\n${buildWorkflowToc(ctx.readProjectFile(".trellis/workflow.md"))}\n</workflow>`)

  const guidelineParts = [
    "These are guideline indexes only. Read the specific files they reference before implementation.",
    "",
  ]
  writeGuidelineIndexes(guidelineParts, ctx, allowedPkgs)
  parts.push(`<guidelines>\n${guidelineParts.join("\n")}\n</guidelines>`)

  parts.push(`<task-status>\n${getTaskStatus(ctx)}\n</task-status>`)

  parts.push(`<ready>\nContext loaded. Workflow index, project state, guideline indexes, and any persisted session gate state are already injected above.\nWait for the user's first message, then follow the workflow.\nIf there is an active task, ask whether to continue it.\n</ready>`)

  return parts.join("\n\n")
}

export default async ({ directory }) => {
  const ctx = new TrellisContext(directory)
  debugLog("session", "Plugin loaded, directory:", directory)

  return {
    "chat.message": async (input) => {
      try {
        const sessionID = input.sessionID
        const agent = input.agent || "unknown"
        debugLog("session", "chat.message called, sessionID:", sessionID, "agent:", agent)

        if (process.env.OPENCODE_NON_INTERACTIVE === "1") {
          debugLog("session", "Skipping - non-interactive mode")
          return
        }

        if (ctx.shouldSkipHook("session-start")) {
          debugLog("session", "Skipping - omo will handle via .claude/hooks/")
          return
        }

        if (contextCollector.isProcessed(sessionID)) {
          debugLog("session", "Skipping - session already processed")
          return
        }

        contextCollector.markProcessed(sessionID)
        const context = buildSessionContext(ctx)
        debugLog("session", "Built context, length:", context.length)
        contextCollector.store(sessionID, context)
        debugLog("session", "Context stored for session:", sessionID)
      } catch (error) {
        debugLog("session", "Error in chat.message:", error.message, error.stack)
      }
    },

    "experimental.chat.messages.transform": async (_input, output) => {
      try {
        const { messages } = output
        debugLog("session", "messages.transform called, messageCount:", messages?.length)

        if (!messages || messages.length === 0) {
          return
        }

        let lastUserMessageIndex = -1
        for (let i = messages.length - 1; i >= 0; i--) {
          if (messages[i].info?.role === "user") {
            lastUserMessageIndex = i
            break
          }
        }

        if (lastUserMessageIndex === -1) {
          debugLog("session", "No user message found")
          return
        }

        const lastUserMessage = messages[lastUserMessageIndex]
        const sessionID = lastUserMessage.info?.sessionID
        debugLog("session", "Found user message, sessionID:", sessionID)

        if (!sessionID || !contextCollector.hasPending(sessionID)) {
          debugLog("session", "No pending context for session")
          return
        }

        const pending = contextCollector.consume(sessionID)
        const textPartIndex = lastUserMessage.parts?.findIndex(
          p => p.type === "text" && p.text !== undefined
        )

        if (textPartIndex === -1) {
          debugLog("session", "No text part found in user message")
          return
        }

        const originalText = lastUserMessage.parts[textPartIndex].text || ""
        lastUserMessage.parts[textPartIndex].text = `${pending.content}\n\n---\n\n${originalText}`
        debugLog("session", "Injected context by prepending to text, length:", pending.content.length)
      } catch (error) {
        debugLog("session", "Error in messages.transform:", error.message, error.stack)
      }
    }
  }
}
